Changes in module 'nf-core/seqtk/subseq'
--- modules/nf-core/seqtk/subseq/main.nf
+++ modules/nf-core/seqtk/subseq/main.nf
@@ -8,11 +8,11 @@
         'quay.io/biocontainers/seqtk:1.3--h5bf99c6_3' }"
 
     input:
-    path sequences
+    tuple val(meta), path sequences
     path filter_list
 
     output:
-    path "*.gz"         , emit: sequences
+    path "*.{fa,fq}"         , emit: sequences
     path "versions.yml" , emit: versions
 
     when:
@@ -30,8 +30,7 @@
         subseq \\
         $args \\
         $sequences \\
-        $filter_list | \\
-        gzip --no-name > ${sequences}${prefix}.${ext}.gz
+        $filter_list > ${sequences}${prefix}.${ext}
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
