Changes in module 'nf-core/seqtk/subseq'
--- modules/nf-core/seqtk/subseq/main.nf
+++ modules/nf-core/seqtk/subseq/main.nf
@@ -12,7 +12,7 @@
     path filter_list
 
     output:
-    path "*.gz"         , emit: sequences
+    path "*.{fa,fq}"         , emit: sequences
     path "versions.yml" , emit: versions
 
     when:
@@ -30,8 +30,7 @@
         subseq \\
         $args \\
         $sequences \\
-        $filter_list | \\
-        gzip --no-name > ${sequences}${prefix}.${ext}.gz
+        $filter_list > ${sequences}${prefix}.${ext}
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
