Changes in module 'nf-core/mitohifi/mitohifi'
--- modules/nf-core/mitohifi/mitohifi/main.nf
+++ modules/nf-core/mitohifi/mitohifi/main.nf
@@ -7,7 +7,7 @@
     container 'ghcr.io/marcelauliano/mitohifi:master'
 
     input:
-    tuple val(meta), path(reads), path(contigs)
+    tuple val(meta), path(input)
     path ref_fa
     path ref_gb
     val mito_code
@@ -41,11 +41,12 @@
     }
 
     def args = task.ext.args ?: ''
-    def run_type = reads ? "-r ${reads}" :
-                    contigs ? "-c ${contigs}" :
-                    exit("Reads or contigs must be specified")
+    def args2 = task.ext.args2 ?: '' 
+    def script_input = args2.equals("-r") ? "-r ${input}" :
+                args2.equals("-c") ? "-c ${input}" :
+                exit("-r for reads or -c for contigs must be specified")
     """
-    mitohifi.py ${run_type} -f ${ref_fa} -g ${ref_gb} -o ${mito_code} -t $task.cpus ${args}
+    mitohifi.py ${script_input} -f ${ref_fa} -g ${ref_gb} -o ${mito_code} -t $task.cpus ${args}
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         mitohifi: \$( mitohifi.py --version 2>&1 | head -n1 | sed 's/^.*MitoHiFi //; s/ .*\$//' )

************************************************************
