Changes in module 'nf-core/mitohifi/findmitoreference'
--- modules/nf-core/mitohifi/findmitoreference/main.nf
+++ modules/nf-core/mitohifi/findmitoreference/main.nf
@@ -1,11 +1,14 @@
 process MITOHIFI_FINDMITOREFERENCE {
+    secret 'NCBI_API_KEY'
+
     tag '$species'
     label 'process_low'
 
     // Docker image available at the biocontainers Dockerhub
-    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'docker://biocontainers/mitohifi:3.0.0_cv1':
-        'docker.io/biocontainers/mitohifi:3.0.0_cv1' }"
+//    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+//        'docker://biocontainers/mitohifi:3.0.0_cv1':
+//        'docker.io/biocontainers/mitohifi:3.0.0_cv1' }"
+    container 'quay.io/sanger-tol/mitohifi:1.0.1'
 
     input:
     val species
@@ -26,7 +29,8 @@
         --species $species \\
         --email $email \\
         --min_length $min_length \\
-        --outfolder .
+        --outfolder . \\
+        --ncbi-api-key \${NCBI_API_KEY}
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
