Changes in module 'nf-core/hifiasm'
--- modules/nf-core/hifiasm/main.nf
+++ modules/nf-core/hifiasm/main.nf
@@ -2,10 +2,11 @@
     tag "$meta.id"
     label 'process_high'
 
-    conda "${moduleDir}/environment.yml"
-    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/hifiasm:0.19.8--h43eeafb_0' :
-        'biocontainers/hifiasm:0.19.8--h43eeafb_0' }"
+    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
+        exit 1, "This version of HIFIASM module does not support Conda. Please use Docker / Singularity / Podman instead."
+    }
+
+    container "wave.seqera.io/wt/73ac3caec075/wave/build:hifiasm-0.19.8_samtools-1.20--1f6824530f0d0ad5"
 
     input:
     tuple val(meta), path(reads)
@@ -13,6 +14,7 @@
     path  maternal_kmer_dump
     path  hic_read1
     path  hic_read2
+    path  hic_reads_cram
 
     output:
     tuple val(meta), path("*.r_utg.gfa")       , emit: raw_unitigs
@@ -23,8 +25,10 @@
     tuple val(meta), path("*.p_utg.gfa")       , emit: processed_unitigs, optional: true
     tuple val(meta), path("*.asm.p_ctg.gfa")   , emit: primary_contigs  , optional: true
     tuple val(meta), path("*.asm.a_ctg.gfa")   , emit: alternate_contigs, optional: true
-    tuple val(meta), path("*.hap1.p_ctg.gfa")  , emit: paternal_contigs , optional: true
-    tuple val(meta), path("*.hap2.p_ctg.gfa")  , emit: maternal_contigs , optional: true
+    tuple val(meta), path("*.asm.hic.p_ctg.gfa")   , emit: hic_primary_contigs  , optional: true
+    tuple val(meta), path("*.asm.hic.a_ctg.gfa")   , emit: hic_alternate_contigs  , optional: true
+    tuple val(meta), path("*.asm.hic.hap1.p_ctg.gfa")  , emit: paternal_contigs , optional: true
+    tuple val(meta), path("*.asm.hic.hap2.p_ctg.gfa")  , emit: maternal_contigs , optional: true
     tuple val(meta), path("*.log")             , emit: log
     path  "versions.yml"                       , emit: versions
 
@@ -34,6 +38,8 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    def hic_read1 = hic_reads_cram ? "for f in $hic_reads_cram; do samtools cat \$f | samtools fastq -n -f0x40 -F0xB00; done" : ""
+    def hic_read2 = hic_reads_cram ? "for f in $hic_reads_cram; do samtools cat \$f | samtools fastq -n -f0x80 -F0xB00; done" : ""
     if ((paternal_kmer_dump) && (maternal_kmer_dump) && (hic_read1) && (hic_read2)) {
         error "Hifiasm Trio-binning and Hi-C integrated should not be used at the same time"
     } else if ((paternal_kmer_dump) && !(maternal_kmer_dump)) {

************************************************************
